File: WPF/Controls/Window.vb
Imports Ks.Common.MVVM

Namespace Controls

    ' ToDo Add some context class that the main window can be get from it. (We kind of have it right now, the KsApplication class is our context class, we just have to make the main objects (like windows and pages and ...) have it. (And maybe it is not needed in this kind of design. Because a view is known only by its view model, and for a view by itself, there is no use.)

    <TemplatePart(Name:=Window.PopupsPanelName, Type:=GetType(PopupPanel))>
    <TemplatePart(Name:=Window.ContentPresenterName, Type:=GetType(ContentPresenter))>
    Public Class Window
        Inherits Windows.Window
        Implements INavigationView

        Shared Sub New()
            DefaultStyleKeyProperty.OverrideMetadata(GetType(Window), New FrameworkPropertyMetadata(GetType(Window)))
            ContentProperty.OverrideMetadata(GetType(Window), New FrameworkPropertyMetadata(Nothing, Nothing, Function(D, C) TryCast(C, UIElement)))
        End Sub

        Friend Sub AddPopup(ByVal Popup As Popup)
            Dim LayerIndex = Popup.Layer
            Dim NextLayerFirst As Popup = Nothing

            For Each KV In Me.ShelterLayers
                If KV.Key > LayerIndex Then
                    NextLayerFirst = KV.Value.Item(0)
                    Exit For
                End If
            Next

            Dim PanelChildren = Me.PopupsPanel.Children
            Dim LastIndex = -1

            If NextLayerFirst Is Nothing Then
                LastIndex = PanelChildren.Count
            Else
                LastIndex = PanelChildren.IndexOf(NextLayerFirst)
            End If

            Dim Layer = Me.ShelterLayers.Item(LayerIndex)
            Dim Shelter As PopupShelter = Nothing

            If Popup.HasShelter Then
                Shelter = New PopupShelter() With {.DataContext = Popup, .Style = Me.ShelterStyle}
                Me.PopupShelterDic.Add(Popup, Shelter)
                PanelChildren.Insert(LastIndex, Shelter)
                LastIndex += 1
            End If

            PanelChildren.Insert(LastIndex, Popup)

            Me.UpdateDims()
        End Sub

        Friend Sub RemovePopup(ByVal Popup As Popup, Optional ByVal LayerIndex As Integer = -1)
            If LayerIndex = -1 Then
                LayerIndex = Popup.Layer
            End If

            Dim Shelter = Me.PopupShelterDic.Item(Popup)
            Me.PopupShelterDic.Remove(Popup)

            Dim PanelChildren = Me.PopupsPanel.Children
            PanelChildren.Remove(Popup)
            PanelChildren.Remove(Shelter)

            Me.UpdateDims()
        End Sub

        Friend Sub RefreshPopup(ByVal Popup As Popup)
            Dim Shelter As PopupShelter = Nothing

            If Popup.HasShelter Then
                If Not Me.PopupShelterDic.ContainsKey(Popup) Then
                    Shelter = New PopupShelter() With {.DataContext = Popup}
                    Me.PopupShelterDic.Add(Popup, Shelter)

                    Dim PanelChildren = Me.PopupsPanel.Children
                    Dim PopupIndex = PanelChildren.IndexOf(Popup)
                    PanelChildren.Insert(PopupIndex, Shelter)
                End If
            Else
                If Me.PopupShelterDic.TryGetValue(Popup, Shelter) Then
                    Me.PopupsPanel.Children.Remove(Shelter)
                    Me.PopupShelterDic.Remove(Popup)
                End If
            End If

            Me.UpdateDims()
        End Sub

        Friend Sub UpdateDims()
            Dim DimmedSeen = False
            For Each I As Integer In Me.ShelterLayers.Keys.Reverse()
                Dim Layer = Me.ShelterLayers.Item(I)
                For J As Integer = Layer.Count - 1 To 0 Step -1
                    Dim Popup = Layer.Item(J)
                    If Popup.HasShelter Then
                        Dim Shelter = Me.PopupShelterDic.Item(Popup)
                        If DimmedSeen Then
                            Shelter.IsShelterShown = False
                        Else
                            Shelter.IsShelterShown = True
                            If Popup.DimShelter Then
                                DimmedSeen = True
                            End If
                        End If
                    End If
                Next
            Next
        End Sub

        Private Sub UpdateShelterStyle()
            Dim ShelterStyle = Me.ShelterStyle
            For Each KV In Me.PopupShelterDic
                KV.Value.Style = ShelterStyle
            Next
        End Sub

        Private Property INavigationView_Content As Page Implements INavigationView.Content
            Get
                Return TryCast(Me.Content, Page)
            End Get
            Set(ByVal Value As Page)
                Me.Content = Value
            End Set
        End Property

#Region "WindowStartupPosition Property"
        Public Shared ReadOnly WindowStartupPositionProperty As DependencyProperty = DependencyProperty.Register("WindowStartupPosition", GetType(WindowStartupLocation), GetType(Window), New PropertyMetadata(WindowStartupLocation.Manual, AddressOf WindowStartupPosition_Changed, AddressOf WindowStartupPosition_Coerce))

        Private Shared Function WindowStartupPosition_Coerce(ByVal D As DependencyObject, ByVal BaseValue As Object) As Object
            'Dim Self = DirectCast(D, Window)

            'Dim Value = DirectCast(BaseValue, WindowStartupLocation)

            Return BaseValue
        End Function

        Private Shared Sub WindowStartupPosition_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, Window)

            'Dim OldValue = DirectCast(E.OldValue, WindowStartupLocation)
            Dim NewValue = DirectCast(E.NewValue, WindowStartupLocation)

            Self.WindowStartupLocation = NewValue
        End Sub

        Public Property WindowStartupPosition As WindowStartupLocation
            Get
                Return DirectCast(Me.GetValue(WindowStartupPositionProperty), WindowStartupLocation)
            End Get
            Set(ByVal value As WindowStartupLocation)
                Me.SetValue(WindowStartupPositionProperty, value)
            End Set
        End Property
#End Region

#Region "ShelterStyle Property"
        Public Shared ReadOnly ShelterStyleProperty As DependencyProperty = DependencyProperty.Register("ShelterStyle", GetType(Style), GetType(Window), New PropertyMetadata(Nothing, AddressOf ShelterStyle_Changed))

        Private Shared Sub ShelterStyle_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, Window)

            'Dim OldValue = DirectCast(E.OldValue, Style)
            'Dim NewValue = DirectCast(E.NewValue, Style)

            Self.UpdateShelterStyle()
        End Sub

        Public Property ShelterStyle As Style
            Get
                Return DirectCast(Me.GetValue(ShelterStyleProperty), Style)
            End Get
            Set(ByVal value As Style)
                Me.SetValue(ShelterStyleProperty, value)
            End Set
        End Property
#End Region

#Region "TemplateParts Logic"
        Public Overrides Sub OnApplyTemplate()
            Me.PopupsPanel = TryCast(Me.Template.FindName(PopupsPanelName, Me), PopupPanel)
            Me.ContentPresenter = TryCast(Me.Template.FindName(ContentPresenterName, Me), ContentPresenter)

            MyBase.OnApplyTemplate()
        End Sub

#Region "PopupsPanel Part"
        Friend Const PopupsPanelName As String = "PART_PopupsPanel"
        Private _PopupsPanel As PopupPanel

        Friend Property PopupsPanel As PopupPanel
            Get
                Return Me._PopupsPanel
            End Get
            Private Set(ByVal Value As PopupPanel)
                Me._PopupsPanel = Value
            End Set
        End Property
#End Region

#Region "ContentPresenter Part"
        Friend Const ContentPresenterName As String = "PART_ContentPresenter"
        Private _ContentPresenter As ContentPresenter

        Private Property ContentPresenter As ContentPresenter
            Get
                Return Me._ContentPresenter
            End Get
            Set(ByVal Value As ContentPresenter)
                Me._ContentPresenter = Value
            End Set
        End Property
#End Region
#End Region

        Private ReadOnly ShelterLayers As CreateInstanceDictionary(Of Integer, List(Of Popup)) = CreateInstanceDictionary.Create(New SortedDictionary(Of Integer, List(Of Popup))())
        Private ReadOnly PopupShelterDic As Dictionary(Of Popup, PopupShelter) = New Dictionary(Of Popup, PopupShelter)()

    End Class

End Namespace

File: WPF/Controls/Window.cs
using System.Windows;
using System.Threading.Tasks;
using Mono;
using System.Data;
using System.Diagnostics;
using Microsoft.VisualBasic;
using System.Linq;
using System.Collections.Generic;
using System.Collections;
using System.Windows.Controls;
using System;
using System.Xml.Linq;
using Ks.Common.MVVM;

namespace Ks
{
    namespace Ks.Common.Controls
    {

        // ToDo Add some context class that the main window can be get from it. (We kind of have it right now, the KsApplication class is our context class, we just have to make the main objects (like windows and pages and ...) have it. (And maybe it is not needed in this kind of design. Because a view is known only by its view model, and for a view by itself, there is no use.)

        [TemplatePart(Name = Window.PopupsPanelName, Type = typeof(PopupPanel))]
        [TemplatePart(Name = Window.ContentPresenterName, Type = typeof(ContentPresenter))]
        public class Window : System.Windows.Window, INavigationView
        {
            static Window()
            {
                DefaultStyleKeyProperty.OverrideMetadata(typeof(Window), new FrameworkPropertyMetadata(typeof(Window)));
                ContentProperty.OverrideMetadata(typeof(Window), new FrameworkPropertyMetadata(null, null, (D, C) => C as UIElement));
            }

            internal void AddPopup(Popup Popup)
            {
                var LayerIndex = Popup.Layer;
                Popup NextLayerFirst = null;

                foreach (var KV in this.ShelterLayers)
                {
                    if (KV.Key > LayerIndex)
                    {
                        NextLayerFirst = KV.Value[0];
                        break;
                    }
                }

                var PanelChildren = this.PopupsPanel.Children;
                var LastIndex = -1;

                if (NextLayerFirst == null)
                    LastIndex = PanelChildren.Count;
                else
                    LastIndex = PanelChildren.IndexOf(NextLayerFirst);

                var Layer = this.ShelterLayers[LayerIndex];
                PopupShelter Shelter = null;

                if (Popup.HasShelter)
                {
                    Shelter = new PopupShelter() { DataContext = Popup, Style = this.ShelterStyle };
                    this.PopupShelterDic.Add(Popup, Shelter);
                    PanelChildren.Insert(LastIndex, Shelter);
                    LastIndex += 1;
                }

                PanelChildren.Insert(LastIndex, Popup);

                this.UpdateDims();
            }

            internal void RemovePopup(Popup Popup, int LayerIndex = -1)
            {
                if (LayerIndex == -1)
                    LayerIndex = Popup.Layer;

                var Shelter = this.PopupShelterDic[Popup];
                this.PopupShelterDic.Remove(Popup);

                var PanelChildren = this.PopupsPanel.Children;
                PanelChildren.Remove(Popup);
                PanelChildren.Remove(Shelter);

                this.UpdateDims();
            }

            internal void RefreshPopup(Popup Popup)
            {
                PopupShelter Shelter = null;

                if (Popup.HasShelter)
                {
                    if (!this.PopupShelterDic.ContainsKey(Popup))
                    {
                        Shelter = new PopupShelter() { DataContext = Popup };
                        this.PopupShelterDic.Add(Popup, Shelter);

                        var PanelChildren = this.PopupsPanel.Children;
                        var PopupIndex = PanelChildren.IndexOf(Popup);
                        PanelChildren.Insert(PopupIndex, Shelter);
                    }
                }
                else if (this.PopupShelterDic.TryGetValue(Popup, out Shelter))
                {
                    this.PopupsPanel.Children.Remove(Shelter);
                    this.PopupShelterDic.Remove(Popup);
                }

                this.UpdateDims();
            }

            internal void UpdateDims()
            {
                var DimmedSeen = false;
                foreach (int I in this.ShelterLayers.Keys.Reverse())
                {
                    var Layer = this.ShelterLayers[I];
                    for (int J = Layer.Count - 1; J >= 0; J += -1)
                    {
                        var Popup = Layer[J];
                        if (Popup.HasShelter)
                        {
                            var Shelter = this.PopupShelterDic[Popup];
                            if (DimmedSeen)
                                Shelter.IsShelterShown = false;
                            else
                            {
                                Shelter.IsShelterShown = true;
                                if (Popup.DimShelter)
                                    DimmedSeen = true;
                            }
                        }
                    }
                }
            }

            private void UpdateShelterStyle()
            {
                var ShelterStyle = this.ShelterStyle;
                foreach (var KV in this.PopupShelterDic)
                    KV.Value.Style = ShelterStyle;
            }

            private Page INavigationView_Content
            {
                get
                {
                    return this.Content as Page;
                }
                set
                {
                    this.Content = value;
                }
            }

            public static readonly DependencyProperty WindowStartupPositionProperty = DependencyProperty.Register("WindowStartupPosition", typeof(WindowStartupLocation), typeof(Window), new PropertyMetadata(WindowStartupLocation.Manual, WindowStartupPosition_Changed, WindowStartupPosition_Coerce));

            private static object WindowStartupPosition_Coerce(DependencyObject D, object BaseValue)
            {
                // Dim Self = DirectCast(D, Window)

                // Dim Value = DirectCast(BaseValue, WindowStartupLocation)

                return BaseValue;
            }

            private static void WindowStartupPosition_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
            {
                var Self = (Window)D;

                // Dim OldValue = DirectCast(E.OldValue, WindowStartupLocation)
                var NewValue = (WindowStartupLocation)E.NewValue;

                Self.WindowStartupLocation = NewValue;
            }

            public WindowStartupLocation WindowStartupPosition
            {
                get
                {
                    return (WindowStartupLocation)this.GetValue(WindowStartupPositionProperty);
                }
                set
                {
                    this.SetValue(WindowStartupPositionProperty, value);
                }
            }

            public static readonly DependencyProperty ShelterStyleProperty = DependencyProperty.Register("ShelterStyle", typeof(Style), typeof(Window), new PropertyMetadata(null, ShelterStyle_Changed));

            private static void ShelterStyle_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
            {
                var Self = (Window)D;

                // Dim OldValue = DirectCast(E.OldValue, Style)
                // Dim NewValue = DirectCast(E.NewValue, Style)

                Self.UpdateShelterStyle();
            }

            public Style ShelterStyle
            {
                get
                {
                    return (Style)this.GetValue(ShelterStyleProperty);
                }
                set
                {
                    this.SetValue(ShelterStyleProperty, value);
                }
            }

            public override void OnApplyTemplate()
            {
                this.PopupsPanel = this.Template.FindName(PopupsPanelName, this) as PopupPanel;
                this.ContentPresenter = this.Template.FindName(ContentPresenterName, this) as ContentPresenter;

                base.OnApplyTemplate();
            }

            internal const string PopupsPanelName = "PART_PopupsPanel";
            private PopupPanel _PopupsPanel;

            internal PopupPanel PopupsPanel
            {
                get
                {
                    return this._PopupsPanel;
                }
                private set
                {
                    this._PopupsPanel = value;
                }
            }

            internal const string ContentPresenterName = "PART_ContentPresenter";
            private ContentPresenter _ContentPresenter;

            private ContentPresenter ContentPresenter
            {
                get
                {
                    return this._ContentPresenter;
                }
                set
                {
                    this._ContentPresenter = value;
                }
            }

            private readonly CreateInstanceDictionary<int, List<Popup>> ShelterLayers = CreateInstanceDictionary.Create(new SortedDictionary<int, List<Popup>>());
            private readonly Dictionary<Popup, PopupShelter> PopupShelterDic = new Dictionary<Popup, PopupShelter>();
        }
    }
}
